Feature/unified configuration file

PhilipJonasFranz opened the pull request:
This PR unifies the `icon_overrides.yml`, `services.yml` files and environment variables into a single configuration file for easier configuration and docker compose setup. A sample for this configuration file is shown in the README and below. In the current implementation, environment variables take precedence over the configuration file, which takes precedence over the default values. The configuration in the server is handled as a unified struct. 

```yaml
selfhstIconUrl: https://cdn.jsdelivr.net/gh/selfhst/icons/
searchUrl: https://www.google.com/search?q=
refreshIntervalSeconds: 30
traefikAPIHost: http://traefik:8080
logLevel: info # info or debug

overrides:
  # Format: <traefik_router_name>: <icon_url_or_filename>

  # Example 1: Using a full URL
  firefly-core: https://selfh.st/content/images/2023/09/favicon-1.png

  # Example 2: Using a filename with .png extension
  unifi-controller: ubiquiti-unifi.png

  # Example 3: Using a filename with .svg extension
  home-assistant: home-assistant.svg

  # Example 4: Using a filename with .webp extension
  plex: plex.webp

exclusions:
  # Format: List router names to exclude from the dashboard

  # Example 1: Exclude the Traefik API router
  - traefik-api
  
  # Example 2: Exclude a private admin interface
  - private-admin-panel
```

This PR also adds a customizable selfhst-icons URL as configuration option and environment variable to be able to specify a base-URL to a self-hosted instance.

PhilipJonasFranz commented:
Marked as draft to get initial feedback.

dannybouwers commented:
> The manual section is a great idea, i was thinking about that too, but i think it should not be in this PR to not blow up the scope of changes.

Agree!

PhilipJonasFranz commented:
I have updated the configuration file structure to your proposed structure, without the manual section. I have moved the Traefik configuration into a separate block to allow extensions in the future.

dannybouwers commented on server/main.go:
```diff
@@ -342,13 +295,13 @@ func findBestIconURL(routerName, serviceURL string) string {
 		ext := filepath.Ext(iconValue)
 		if ext == ".png" || ext == ".svg" || ext == ".webp" {
-			url := "https://cdn.jsdelivr.net/gh/selfhst/icons/" + strings.TrimPrefix(ext, ".") + "/" + strings.ToLower(iconValue)
+			url := configuration.EnvironmentConfig.SearchURL + strings.TrimPrefix(ext, ".") + "/" + strings.ToLower(iconValue)
```
- [ ] This should be the SelfhstIconUrl instead of the SearchURL.
- [ ] This could result in malformed URLs if SearchURL doesn't end with a `/`. we need logic to check this and add a `/` if necessary.

dannybouwers commented on server/main.go:
```diff
@@ -342,13 +295,13 @@ func findBestIconURL(routerName, serviceURL string) string {
 
 		// Fallback to default behavior if extension is not valid
-		url := "https://cdn.jsdelivr.net/gh/selfhst/icons/png/" + iconValue
+		url := configuration.EnvironmentConfig.SearchURL + "png/" + iconValue
```
- [ ] This should be the SelfhstIconUrl instead of the SearchURL.
- [ ] This could result in malformed URLs if SearchURL doesn't end with a `/`. we need logic to check this and add a `/` if necessary.

dannybouwers commented on README.md:
```diff
@@ -75,119 +73,77 @@ networks:
 
-3. Mount this file into the container at `/config/icon_overrides.yml` using a volume, as shown in the `docker-compose.yml` example.
+environment:
+  selfhstIconUrl: https://cdn.jsdelivr.net/gh/selfhst/icons/
```
There's an inconsistency in naming between the README and code:

 * README shows selfhstIconUrl in the config file
 * Code uses SelfhstIconUrl (camelCase vs PascalCase)
 * Environment variable is SELFHST_ICON_URL
 
 Docker uses snake case (all lowercase and underscores between words), so I suggest using that convention in TraLa config as well.

dannybouwers commented on server/main.go:
```diff
@@ -622,12 +576,99 @@ func resolveURL(baseURL string, path string) (string, error) {
+	if config.EnvironmentConfig.LogLevel == "debug" {
+		debugf("Using effective configuration:")
+		out, err := yaml.Marshal(config)
+		if err != nil {
```
Great addition to lof the active configuration in debug mode. However, if this error occurs on the configuration the application just constructed, wouldn't we have bigger issues in the application in general? Could you fill me in there?

dannybouwers reviewed (changes requested):
The PR is well-implemented and improves the application's configuration management. Thanks al lot! However, I recommend addressing two issues before merging:

- [ ] Fix the URL construction bug in findBestIconURL
- [ ] Naming inconsistency between README and code for selfhst icon URL configuration. I suggest using snake_case in the config file to align with both docker standards and the environment variables.

I've also added a question in the comments. I hope you can provide some more insights...
